nano/node/transport/tcp.cpp

```diff
@@ -63,3 +71,561 @@ std::string nano::transport::channel_tcp::to_string () const
+{
+	if (!max_ip_connections (endpoint_a))
+	{
+		// Check for special node port data
+		std::vector<nano::tcp_endpoint> insert_response_channels;
+		auto peer0 (message_a.peers[0]);
+		auto peer1 (message_a.peers[1]);
+		if (peer0.address () == boost::asio::ip::address_v6{} && peer0.port () != 0)
+		{
+			nano::endpoint new_endpoint (endpoint_a.address (), peer0.port ());
+			node.network.merge_peer (new_endpoint);
+			if (keepalive_first)
+			{
+				insert_response_channels.push_back (nano::transport::map_endpoint_to_tcp (new_endpoint));
+			}
+		}
+		if (peer1.address () != boost::asio::ip::address_v6{} && peer1.port () != 0 && keepalive_first)
+		{
+			insert_response_channels.push_back (nano::transport::map_endpoint_to_tcp (peer1));
+		}
+		// Insert preferred response channels from first TCP keepalive
+		if (!insert_response_channels.empty ())
+		{
+			node.network.add_response_channels (endpoint_a, insert_response_channels);
+		}
+		auto udp_channel (std::make_shared<nano::transport::channel_udp> (node.network.udp_channels, nano::transport::map_tcp_to_endpoint (endpoint_a)));
+		node.process_message (message_a, udp_channel);
+	}
+}
+
+void nano::transport::tcp_channels::start ()
+{
+	ongoing_keepalive ();
+	ongoing_syn_cookie_cleanup ();
+}
+
+void nano::transport::tcp_channels::stop ()
+{
+	stopped = true;
+	// Close all TCP sockets
+	for (auto i (channels.begin ()), j (channels.end ()); i != j; ++i)
+	{
+		if (i->channel->socket != nullptr)
+		{
+			i->channel->socket->close ();
+		}
+	}
+}
+
+bool nano::transport::tcp_channels::max_ip_connections (nano::tcp_endpoint const & endpoint_a)
+{
+	std::unique_lock<std::mutex> lock (mutex);
+	bool result (channels.get<ip_address_tag> ().count (endpoint_a.address ()) >= nano::transport::max_peers_per_ip);
+	return result;
+}
+
+bool nano::transport::tcp_channels::reachout (nano::endpoint const & endpoint_a)
+{
+	auto tcp_endpoint (nano::transport::map_endpoint_to_tcp (endpoint_a));
+	// Don't overload single IP
+	bool error = max_ip_connections (tcp_endpoint);
+	if (!error)
+	{
+		// Don't keepalive to nodes that already sent us something
+		error |= find_channel (tcp_endpoint) != nullptr;
+		std::lock_guard<std::mutex> lock (mutex);
+		auto existing (attempts.find (tcp_endpoint));
+		error |= existing != attempts.end ();
+		attempts.insert ({ tcp_endpoint, std::chrono::steady_clock::now () });
+	}
+	return error;
+}
+
+std::unique_ptr<nano::seq_con_info_component> nano::transport::tcp_channels::collect_seq_con_info (std::string const & name)
+{
+	size_t channels_count = 0;
+	size_t attemps_count = 0;
+	size_t syn_cookies_count = 0;
+	size_t syn_cookies_per_ip_count = 0;
+	{
+		std::lock_guard<std::mutex> guard (mutex);
+		channels_count = channels.size ();
+		attemps_count = attempts.size ();
+		std::lock_guard<std::mutex> syn_cookie_guard (syn_cookie_mutex);
```

Comment:
The first guard can be put inside a new scope to release earlier (and prevent possible deadlock).

Reply:
makes sense

Reply:
Updated